packages/vuetify/src/composables/validation.ts

```diff
@@ -0,0 +1,80 @@
+// Utilities
+import { propsFactory, wrapInPromise } from '@/util'
+import { ref, unref, watch } from 'vue'
+
+// Types
+import type { MaybeRef } from '@/util/helpers'
+import type { PropType } from 'vue'
+
+export type ValidationRule = string | ((value: any) => string | boolean)
+export type ValidationRules = ValidationRule[]
+export type ValidationResult = string | boolean | Promise<string | boolean>
+
+export interface ValidationProps {
+  errorCount: string | number
+  rules: ValidationRules
+}
+
+export const makeValidationProps = propsFactory({
+  errorCount: {
+    type: [Number, String],
+    default: 1,
+  },
+  rules: {
+    type: Array as PropType<ValidationRules>,
+    default: () => ([]),
+  },
+})
+
+export function useValidation (props: ValidationProps, value: MaybeRef<any>) {
+  const errorMessages = ref<string[]>([])
+  const isPristine = ref(true)
+  const isValid = ref<null | boolean>(null)
+  const isValidating = ref(false)
+  const errorCount = ref(0)
+  const unwrefed = unref(value)
+
+  watch(isValid, () => isPristine.value = false)
+
+  async function validate () {
+    await reset()
+
+    isValidating.value = true
+    for (const rule of props.rules) {
+      const handler = typeof rule === 'function' ? rule : () => rule
+      const valid = await wrapInPromise<ValidationResult>(handler(unwrefed.value))
+
+      if (valid === true) continue
+
+      errorCount.value += 1
+
+      if (typeof valid === 'string') {
+        errorMessages.value.push(valid)
+      }
+
+      if (errorCount.value >= (props.errorCount || 1)) {
+        break
+      }
+    }
+    isValid.value = errorCount.value > 0
```

Comment:
?

Reply:
errorCount !== props.errorCount. Not all errors get pushed into messages so every error iterates the total number of errors (which is used for determining the rules validation loop breaks). This will be less confusing with the suggested prop name change you made.

Reply:
Then why set `isValid.value` again two lines down?

Anyway, now I understand the purpose of `errorCount` variable, but I don't think I agree with it. As I mentioned here https://github.com/vuetifyjs/vuetify/pull/14026/files#r683723806 I think it is a bad idea to silently ignore rules that don't have a string.

IMO we should be nice to developers here (and by extension end users) by warning that a rule is missing a string message. If developer really really want to fail an input without a visible message, they can do `v => !!v || ''`

Reply:
The second isValid comparison was accidentally left in.